compiler/rustc_hir/src/lang_items.rs

```diff
@@ -273,6 +273,7 @@ language_item_table! {
     EhCatchTypeinfo,         sym::eh_catch_typeinfo,   eh_catch_typeinfo,          Target::Static,         GenericRequirement::None;
 
     OwnedBox,                sym::owned_box,           owned_box,                  Target::Struct,         GenericRequirement::Minimum(1);
+    OwnedBoxNew,             sym::owned_box_new,       owned_box_new,              Target::Method(MethodKind::Inherent), GenericRequirement::None;
```

Comment:
Making it a diagnostic item instead of a lang item would be preferable in this case.

Reply:
I've been planning, as a followup to this PR, to also use this for the `#[rustc_box]` feature. Right now, `#[rustc_box] foobar(foo)` is lowered to `box foo`, irrespective of what `foobar` is, even if it isn't `Box::new`. In order to make it error if it's not `Box::new`, I would be needing something that gives me a def id i suppose. Is a diagnostic item a good choice for such an error as well?

To explain, in a `box_syntax` free world, some users might still want to use `#[rustc_box]` over `Box::new` and they should get a more polished feature.

Reply:
>Right now, `#[rustc_box] foobar(foo)` is lowered to `box foo`, irrespective of what `foobar` is, even if it isn't `Box::new`.

Which seems fine to me, given that `#[rustc_box]` is an internal feature that is supposed to be removed in the future.

>free world, some users might still want to use `#[rustc_box]` over `Box::new` and they should get a more polished feature.

If it's supposed to be used by people and be polished, then perhaps it needs a better syntax than `#[rustc_box] Box::new(expr)`, for example `box expr`, oh wait.

Reply:
I didn't follow the story behind https://github.com/rust-lang/rust/pull/97293 but the state in which you need to write `#[rustc_box] Box::new(expr)` instead of just `Box::new(expr)` or `box expr` (*) seems like the worse from both worlds.

Preferably the former, because whatever made `Box::new(expr)` slow in https://github.com/rust-lang/rust/pull/97293 can also make other much more common things like `Vec::from()`/`String::from()` slow, so it may be a good win if it's fixed.

Reply:
`Box::new(foo)` is not slow in *most* of cases, thanks to LLVM being really good, especially [post version 12](https://github.com/rust-lang/rust/pull/87781#issuecomment-894580531) (minimum LLVM is 12 now). The issue is just that switching `vec![]` to `Box::new` would also switch it for the remaining few cases where it does cause regressions. For these, the PR added `#[rustc_box]` as an alternative to `box foo`. My ultimate goal is to remove `box foo` support from the compiler, this PR is another preparation of it. Eventually this will need a lang team FCP and a "go forward" from them. Anyways, the general future of `box` syntax is not discussed here but in the tracking issue (still need to make a proposal there).

It might be better if instead of doing empty promises, I add the lang item in a separate PR first that actually needs it, and then switch this PR to it.